src/ui/FacetValueSuggestions/FacetValueSuggestionsProvider.ts

```diff
@@ -0,0 +1,179 @@
+
+      const suggestionRows = filteredResults.map(indexFieldValue => {
+        return <IFacetValueSuggestionRow>{
+          numberOfResults: indexFieldValue.numberOfResults,
+          keyword: fieldResponse.keyword,
+          value: indexFieldValue.value,
+          score: this.computeScoreForSuggestionRow(indexFieldValue, fieldTotalReference),
+          field: this.options.field
+        };
+      });
+      return allValues.concat(suggestionRows);
+    }, []);
+
+    return resultsWithoutSelectedValues;
+  }
+
+  private async getFieldValuesToQuery(valuesToSearch: string[]): Promise<IFacetValueSuggestionsResponse> {
+    // The reference request will be used to get the maximum number of values for a given facet value.
+    const referenceValuesRequest = this.buildReferenceFieldValueRequest();
+    const queryParts = this.getQueryToExecuteParts();
+    const suggestionValuesRequests = valuesToSearch.map(value => this.buildListFieldValueRequest(queryParts.concat(value).join(' ')));
+    const requests = [].concat(suggestionValuesRequests).concat(referenceValuesRequest);
+    const values = await this.queryController.getEndpoint().listFieldValuesBatch({
+      batch: requests
+    });
+
+    const reference = this.computeReferenceFromBatch(values.pop());
+    const responses: IFacetValueBatchResponse[] = values.map((value, i) => {
+      return <IFacetValueBatchResponse>{
+        keyword: valuesToSearch[i],
+        values: value
+      };
+    });
+
+    return <IFacetValueSuggestionsResponse>{
+      responses,
+      reference
+    };
+  }
+
+  private computeScoreForSuggestionRow(fieldValue: IIndexFieldValue, reference: IFacetValueReference): IFacetValueSuggestionScore {
+    const totalNumberForFieldValue = reference.fieldsTotal[fieldValue.value] || reference.smallestTotal;
+    const distanceFromTotalForField: number = (totalNumberForFieldValue - fieldValue.numberOfResults) / totalNumberForFieldValue * 100;
+    return {
+      distanceFromTotalForField
+    };
+  }
+
+  private computeReferenceFromBatch(batch: IIndexFieldValue[]): IFacetValueReference {
+    const fieldsTotal = {};
+    batch.forEach(value => (fieldsTotal[value.value] = value.numberOfResults));
+    return {
+      fieldsTotal: fieldsTotal,
+      smallestTotal: batch[batch.length - 1].numberOfResults
+    };
+  }
+
+  private buildListFieldValueRequest(queryToExecute: string): IListFieldValuesRequest {
+    return {
+      field: <string>this.options.field,
+      ignoreAccents: true,
+      maximumNumberOfValues: 3,
+      queryOverride: queryToExecute
+    };
+  }
+
+  private buildReferenceFieldValueRequest(): IListFieldValuesRequest {
+    return {
+      field: <string>this.options.field
+    };
+  }
+
+  private getQueryToExecuteParts(): string[] {
+    const lastQuery = this.queryController.getLastQuery();
+    const aqWithoutCurrentField =
```

Comment:
Could use `QueryBuilder.computeCompleteExpressionExcept` ? It's the method used by standard facets

Reply:
Hm, I have mixed feelings on this suggestions :P

I'm not in a QueryBuilder context, only on a `getLastQuery`, which means the parts are already merged together and I would rather not try to split them.